src/sage/categories/coxeter_groups.py

```diff
@@ -1971,9 +1975,125 @@ def absolute_length(self):
+
+                sage: W = CoxeterGroup(["A",2,1])
+                sage: (r, s, t) = W.simple_reflections()
+                sage: (r * s * r * t).absolute_length()
+                2
+                sage: W.one().absolute_length()
+                0
+                sage: r.absolute_length()
+                1
+                sage: (r * s).absolute_length()
+                2
+                sage: (r * s * r).absolute_length()
+                1
             """
-            M = self.canonical_matrix()
-            return (M - 1).image().dimension()
+            P = self.parent()
+            if P.is_finite():
+                M = self.canonical_matrix()
+                return (M - 1).image().dimension()
+
+            n = self.length()
+            if n <= 2:  # trivial cases
+                return n
+            return len(self.absolute_chain_reflections())
+
+        def absolute_chain(self):
+            r"""
+            Return a (saturated) chain in absolute order from ``self``
+            to ``1``.
+
+            .. SEEALSO::
+
+                :meth:`absolute_chain_reflections`
+
+            EXAMPLES::
+
+                sage: W = CoxeterGroup(['A', 2, 1])
+                sage: (r, s, t) = W.simple_reflections()
+                sage: (r * s * r * t).absolute_chain()
+                [
+                [ 2  1 -2]  [ 0 -1  2]  [1 0 0]
+                [ 1  2 -2]  [-1  0  2]  [0 1 0]
+                [ 1  1 -1], [ 0  0  1], [0 0 1]
+                ]
+            """
+
+            reflections = self.absolute_chain_reflections()
+            P = self.parent()
+            return [P.prod(reversed(reflections[:i]))
+                    for i in range(len(reflections), -1, -1)]
+
+
+        def absolute_chain_reflections(self):
+            r"""
+            Return a list of reflection which, when multiplied in order, give
+            ``self``.
+
+
+            This method is based on Theorem 1.1 in [Dy2001]_, combined with
+            the strong exchange condition.
+
+            .. SEEALSO::
+
+                :meth:`absolute_length`
+                :meth:`absolute_chain`
+
+            EXAMPLES::
+
+                sage: W = CoxeterGroup(["A",2,1])
+                sage: W.one().absolute_chain_reflections()
+                []
+                sage: (r, s, t) = W.simple_reflections()
+                sage: r.absolute_chain_reflections()
+                [
+                [-1  1  1]
+                [ 0  1  0]
+                [ 0  0  1]
+                ]
+                sage: (r * s).absolute_chain_reflections()
+                [
+                [-1  1  1]  [ 0 -1  2]
+                [ 0  1  0]  [-1  0  2]
+                [ 0  0  1], [ 0  0  1]
+                ]
+                sage: (r * s * r * t).absolute_chain_reflections()
+                [
+                [ 0 -1  2]  [-1 -2  4]
+                [-1  0  2]  [-2 -1  4]
+                [ 0  0  1], [-1 -1  3]
+                ]
+
+            """
```

Comment:
```suggestion
                ]
            """
```
This is for consistency with the rest of Sage (at least, what we are moving towards).